contributors/design-proposals/bulk_watch.md

```diff
@@ -0,0 +1,456 @@
+responsible for aggregating requests from underlying frameworks into a single
+one and then dispatching incoming watch events to correct reflect/informer.
+This will obviously require changes to existing frameworks, but those should
+be local changes.
+
+ One more thing to mention is detecting resource type of object being send via
+watch. With bulk watch, you don't really know object of what type has just been
+sent to you.
+
+ To address that, I suggest introducing a dedicated type for bulk watch events
+that will additionally contain resource type of the object it contains. It will
+look as following:
+```
+type BulkEvent struct {
+	Resource GroupVersionResource
+
+	Type   watch.EventType
+	Object runtime.Object
+}
+```
+
+*TODO: We already have some logic to detect object type by deserializing only
+parts of it. It may be possible to reuse it instead, but we would need to prove
+that it works now (also for protobufs).*
+
+
+### Dynamic watch
+
+ As mentioned in the Proposal section, we will implement bulk watch that will
+allow for dynamic subscribtion/unsubscribtion for (sets of) objects on top of
+websockets protocol.
+
+ Note that we already support websockets in the regular Kubernetes API for
+watch requests (in addition to regular http requests), so for the purpose of
+bulk watch we will be extending websocket support.
+
+ The the high level, the propocol will look:
+1. client opens a new websocket connection to a bulk watch endpoint to the
+server via gttp GET
+1. this results in creating in creating a single channel that is used only
+to handle communication for subscribing/unsubscribing for watches; no watch
+events are delivered via this particular channel.  
+*TODO: Consider switching to two channels, one for incoming and one for
+outgoing communication*
+1. to subscribe for a watch of a given (set of) objects, user sends `Subscribe`
+object over the channel; in response a new channel is created and the message
+with the channel identifier is send back to the user (we will be using integers
+as channel identifiers).
+1. once subscribed, all objects matching a given selector will be send over
+the newly created channel
+1. to stop watching for a given (set of) objects, user sends `Unsubscribe`
+object over the the channel; in response the corresponding watch is broken and
+corresponding channel within websocket is closed
+1. once done, user can close the whole websocket connection (this results in
+breaking all still opened channels and corresponding watches).
+
+ With that high level protocol, there are still multiple details that needs
+to be figured out. First, we need to define `Subscribe` and `Unsubscribe`
+message. We will solve it with the single `Request` object:
+```
+type Request struct {
+	// Only one of those is set.
+	Subscribe   Subscribe
+	Unsubscribe Unsubscribe
+}
+
+type Identifier int64
+
+// Watch request for objects matching a given selector.
+type Subscribe struct {
```

Comment:
I meant, while this model looks like publish subscribe, rather than use the exact wording, we use a wording that is a bit more in line with our existing clients and with gRPC than one is more in line with publish./subscribe.  Our client verb is Watch, and then you Close() the watch if you're done.  So we want to have an underlying protocol that mimic thats that:

* Watch starts a watch
* WatchClose/CloseWatch closes a watch